cpp/src/io/utilities/datasource.cpp

```diff
@@ -199,6 +199,47 @@ class direct_read_source : public file_source {
   }
 };
 
+/**
+ * @brief Implementation class for reading from a device buffer source
+ */
+class device_buffer_source : public datasource {
+ public:
+  explicit device_buffer_source(cudf::device_span<uint8_t> d_buffer) : _d_buffer{d_buffer} {}
+
+  size_t host_read(size_t offset, size_t size, uint8_t* dst) final
+  {
+    CUDF_FAIL("Host read shouldn't be used with device_buffer_source");
+  }
+
+  std::unique_ptr<buffer> host_read(size_t offset, size_t size) final
+  {
+    CUDF_FAIL("Host read shouldn't be used with device_buffer_source");
+  }
+
+  [[nodiscard]] bool supports_device_read() const final { return true; }
+
+  size_t device_read(size_t offset, size_t size, uint8_t* dst, rmm::cuda_stream_view stream) final
```

Comment:
You have to implement `device_read_async` as well.
This one's a bit tricky. I'm trying to figure out if we can just do an async D2D copy and return a finished task.